ArcGISRuntimeSDKQt_CppSamples/Scenes/Add3DTilesLayer/Add3DTilesLayer.cpp

```diff
@@ -0,0 +1,211 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#ifdef PCH_BUILD
+#include "pch.hpp"
+#endif // PCH_BUILD
+
+#include "Add3DTilesLayer.h"
+
+#include "ArcGISTiledElevationSource.h"
+#include "ElevationSourceListModel.h"
+#include "MapTypes.h"
+#include "Scene.h"
+#include "SceneQuickView.h"
+#include "Surface.h"
+#include "Basemap.h"
+#include "Ogc3dTilesLayer.h"
+#include "LayerListModel.h"
+#include "Layer.h"
+#include "SpatialReference.h"
+#include "Point.h"
+#include "LocationLineOfSight.h"
+#include "AnalysisOverlay.h"
+#include "AnalysisOverlayListModel.h"
+#include "AnalysisListModel.h"
+#include "Camera.h"
+#include "TransformationCatalog.h"
+
+#include <MapView.h>
+#include <QString>
+#include <QFuture>
+#include <QDebug>
+#include <QStandardPaths>
+#include <QVariantMap>
+#include <QUrl>
+
+using namespace Esri::ArcGISRuntime;
+namespace
+{
+QString defaultDataPath()
+{
+  QString dataPath;
+
+#ifdef Q_OS_IOS
+  dataPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
+#else
+  dataPath = QStandardPaths::writableLocation(QStandardPaths::HomeLocation);
+#endif
+
+  return dataPath;
+}
+}
+
+const QString DATAPATH = QStandardPaths::writableLocation(QStandardPaths::HomeLocation)+ "/ArcGIS/Runtime/Data/";
+
+
+Add3DTilesLayer::Add3DTilesLayer(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_scene(new Scene(BasemapStyle::ArcGISImagery, this))
+{
+  // create a new elevation source from Terrain3D REST service
+  ArcGISTiledElevationSource* elevationSource = new ArcGISTiledElevationSource(
+        QUrl("https://elevation3d.arcgis.com/arcgis/rest/services/WorldElevation3D/Terrain3D/ImageServer"), this);
+  // add projection
+  TransformationCatalog::setProjectionEngineDirectory(DATAPATH + "pedata");
+  if (TransformationCatalog::projectionEngineDirectory().isEmpty())
+    logErrorText("Projection Engine Directory not found at: " + DATAPATH + "pedata");
+
+  // add the elevation source to the scene to display elevation
+  m_scene->baseSurface()->elevationSources()->append(elevationSource);
+
+  // add 3d layers
+  add3DTilesLayer();
+}
+
+Add3DTilesLayer::~Add3DTilesLayer() = default;
+
+void Add3DTilesLayer::init()
+{
+  // Register classes for QML
+  qmlRegisterType<SceneQuickView>("Esri.Samples", 1, 0, "SceneView");
+  qmlRegisterType<Add3DTilesLayer>("Esri.Samples", 1, 0, "Add3DTilesLayerSample");
+}
+
+SceneQuickView* Add3DTilesLayer::sceneView() const
+{
+  return m_sceneView;
+}
+
+// Set the view (created in QML)
+void Add3DTilesLayer::setSceneView(SceneQuickView* sceneView)
+{
+  if (!sceneView || sceneView == m_sceneView)
+    return;
+
+  m_sceneView = sceneView;
+  m_sceneView->setArcGISScene(m_scene);
+
+  emit sceneViewChanged();
+
+  if (!defaultDataPath().isEmpty() && m_sceneView)
+    initialize();
```

Comment:
I'm not sure if m_sceneView would ever be null here, because we check if sceneView is null above, so that check can probably be removed.

Furthermore defaultDataPath is some form of QStandardPaths::writableLocation() which always returns a value for the parameters we gave the method, so we can assume that it will not be empty.

Therefore we can probably remove this condition altogether.
```suggestion
  initialize();
```

Reply:
For `HomeLocation` this is the doc: 
> Returns the user's home directory (the same as [QDir::homePath](https://doc.qt.io/qt-6/qdir.html#homePath)()). On Unix systems, this is equal to the HOME environment variable. This value might be generic or application-specific, but the returned path is never empty.
And for `DocumentsLocation` here's the doc:
> Returns the directory containing user document files. This is a generic value. The returned path is never empty.

Some enums may return empty values that writableLocation can't resolve, but these ones don't.

Reply:
Ah, I see. I'll remove the condition. Thanks for clarifying!